src/snowflake/snowpark/column.py

```diff
@@ -225,108 +238,330 @@ class Column:
+    #       Currently a breaking example can be created using the Column.isin method as it does not build the AST.
+    #       For example, running: df.filter(col("A").isin(1, 2, 3) & col("B")) would fail since the boolean operator
+    #       '&' would try to construct an AST using that of the new col("A").isin(1, 2, 3) column (which we currently
+    #       don't fill if the only argument provided in the Column constructor is 'expr1' of type Expression)
     def __init__(
-        self, expr1: Union[str, Expression], expr2: Optional[str] = None
+        self,
+        expr1: Union[str, Expression],
+        expr2: Optional[str] = None,
+        ast: Optional[proto.Expr] = None,
+        _emit_ast: bool = True,
     ) -> None:
+        self._ast = ast
+
         if expr2 is not None:
-            if isinstance(expr1, str) and isinstance(expr2, str):
-                if expr2 == "*":
-                    self._expression = Star([], df_alias=expr1)
-                else:
-                    self._expression = UnresolvedAttribute(
-                        quote_name(expr2), df_alias=expr1
-                    )
-            else:
+            if not (isinstance(expr1, str) and isinstance(expr2, str)):
                 raise ValueError(
                     "When Column constructor gets two arguments, both need to be <str>"
                 )
+
+            if expr2 == "*":
+                self._expression = Star([], df_alias=expr1)
+            else:
+                self._expression = UnresolvedAttribute(
+                    quote_name(expr2), df_alias=expr1
+                )
+
+            # Alias field should be from the parameter provided to DataFrame.alias(self, name: str)
+            # A column from the aliased DataFrame instance can be created using this alias like col(<df_alias>, <col_name>)
+            # In the IR we will need to store this alias to resolve which DataFrame instance the user is referring to
+            if self._ast is None and _emit_ast:
+                self._ast = create_ast_for_column(expr1, expr2)
+
         elif isinstance(expr1, str):
             if expr1 == "*":
                 self._expression = Star([])
             else:
                 self._expression = UnresolvedAttribute(quote_name(expr1))
+
+            if self._ast is None and _emit_ast:
+                self._ast = create_ast_for_column(expr1, None)
+
         elif isinstance(expr1, Expression):
             self._expression = expr1
+
+            if self._ast is None and _emit_ast:
+                if hasattr(expr1, "_ast"):
+                    self._ast = expr1._ast
+                else:
+                    self._ast = snowpark_expression_to_ast(expr1)
+
         else:  # pragma: no cover
             raise TypeError("Column constructor only accepts str or expression.")
 
+        assert self._expression is not None
+
+    def __emit_ast_for_binary(self, other: Any) -> bool:
```

Comment:
should we call it something like `__should_emit_ast` based on the functionality?

Reply:
Based on the implementation it seems to checking if self and other are `Column` with an AST. maybe updating the docstring is better here?

Reply:
addressed in https://github.com/snowflakedb/snowpark-python/pull/2507/files